File: R/CIR/getMErrors.R
# names must be a factor of strings
# momth is just needed if setup is C (conditonal) or UC AND approach == 3
# Need function MomError for UC settings

getCIRMErrors = function(v0, kappa,theta, sigma, T, n, M, NRuns, names, nE, nI, nO, setup, approach){
  
  Results = list()
  no.cores = detectCores() - 1
  
  if(setup == 'C'){
    
    Evp    = MomentsCIR(p = 1:4, q = 0, kappa = kappa, theta = theta, sigma = sigma, v0 = v0, t = T, conditional = TRUE, comoments = FALSE)
    varT   = Evp[2] - Evp[1]^2
    skewT  = getSkFromMoments(Evp[1], Evp[2], Evp[3])
    kurtT  = getKuFromMoments(Evp[1], Evp[2], Evp[3], Evp[4])
    
    momth = cbind(Evp[1], varT, skewT, kurtT)
    
    Errors = list()
    
    if(no.cores > length(names)){
      
      cl = makeCluster(length(names))
      
    }else{
      
      cl = makeCluster(no.cores)
      
    }
    
    registerDoParallel(cl)
    clusterExport(cl,ls(globalenv()))
    
    for(k in 1:NRuns){
      
      for(j in 1:length(n)){
        
        Errors[[j]] = foreach(i = 1:length(names), .combine = 'cbind') %dopar% TreatVariance(Model = names[i], v0 = v0, kappa = kappa, theta = theta, sigma = sigma, T = T, n = n[j], M = M, momth = momth, setup = setup)
        
      }
      
      Results[[k]] = Errors
      
      print(k)
      
    }
    
    stopCluster(cl)
    
  }else if(setup == 'UC'){
    
    if(approach == 1){ # fix n to 1, change T/Delta. v0 is a Gamma random variate here 
      
      # Plug in dummy for v0 and T since the unconditional moments are not functions of these two inputs
      Evp    = MomentsCIR(p = 1:4, q = 0, kappa = kappa, theta = theta, sigma = sigma, v0 = 1, t = 1, conditional = FALSE, comoments = FALSE)
      varT   = Evp[2] - Evp[1]^2
      skewT  = getSkFromMoments(Evp[1], Evp[2], Evp[3])
      kurtT  = getKuFromMoments(Evp[1], Evp[2], Evp[3], Evp[4])
      
      momth = cbind(Evp[1], varT, skewT, kurtT)
      
      Errors = list()
      
      
      if(no.cores > length(T)){
        
        cl = makeCluster(length(T))
        
      }else{
        
        cl = makeCluster(no.cores)
        
      }
      
      registerDoParallel(cl)
      clusterExport(cl,ls(globalenv()))
      
      for(k in 1:NRuns){
        
        for(j in 1:length(T)){
          
          # To set setup = 'C' here is not a mistake, since we are interested in the distribution of v_T again
          Errors[[j]] = foreach(i = 1:length(names), .combine = 'cbind') %dopar% TreatVariance(Model = names[i], v0 = v0, kappa = kappa, theta = theta, sigma = sigma, T = T[j], n = n, M = M, momth = momth, setup = 'C')
          
        }
        
        Results[[k]] = Errors
        
        print(k)
        
      }
      
      stopCluster(cl)
    
    }else if(approach == 2){ # Fix n to a high number, change T/Delta
      
      # plug in some dummy for T here, since the unconditional moments are no function of time
      Evp    = MomentsCIR(p = 1:4, q = 0, kappa = kappa, theta = theta, sigma = sigma, v0 = v0, t = 1, conditional = FALSE, comoments = FALSE)
      varT   = Evp[2] - Evp[1]^2
      skewT  = getSkFromMoments(Evp[1], Evp[2], Evp[3])
      kurtT  = getKuFromMoments(Evp[1], Evp[2], Evp[3], Evp[4])
      
      momth = cbind(Evp[1], varT, skewT, kurtT)
      
      Errors = list()
      
      if(no.cores > length(names)){
        
        cl = makeCluster(length(names))
        
      }else{
        
        cl = makeCluster(no.cores)
        
      }
      
      registerDoParallel(cl)
      clusterExport(cl,ls(globalenv()))
      
      for(k in 1:NRuns){
        
        for(j in 1:length(T)){
          
          Errors[[j]] = foreach(i = 1:length(names), .combine = 'cbind', .packages = "EnvStats") %dopar% TreatVariance(Model = names[i], v0 = v0, kappa = kappa, theta = theta, sigma = sigma, T = T[j], n = n, M = M, momth = momth, setup = setup)
          
        }
        
        Results[[k]] = Errors
        print(k)
        
      }
      
      stopCluster(cl)
        
    }else if(approach == 3){ # Fix T, change n/Delta 
      
      Evp    = MomentsCIR(p = 1:4, q = 0, kappa = kappa, theta = theta, sigma = sigma, v0 = v0, t = T, conditional = FALSE, comoments = FALSE)
      varT   = Evp[2] - Evp[1]^2
      skewT  = getSkFromMoments(Evp[1], Evp[2], Evp[3])
      kurtT  = getKuFromMoments(Evp[1], Evp[2], Evp[3], Evp[4])
      
      momth = cbind(Evp[1], varT, skewT, kurtT)
      
      Errors = list()
      
      if(no.cores > length(names)){
        
        cl = makeCluster(length(names))
        
      }else{
        
        cl = makeCluster(no.cores)
        
      }
      
      registerDoParallel(cl)
      clusterExport(cl,ls(globalenv()))
      
      for(k in 1:NRuns){
        
        for(j in 1:length(n)){
          
          Errors[[j]] = foreach(i = 1:length(names), .combine = 'cbind') %dopar% TreatVariance(Model = names[i], v0 = v0, kappa = kappa, theta = theta, sigma = sigma, T = T, n = n[j], M = M, momth = momth, setup = setup)
          
        }
        
        Results[[k]] = Errors
        print(k)
        
      }
      
      stopCluster(cl)
      
    }
    
  }
  
  Bsmean = list() 
  Bsvar  = list()
  Bsskew = list()
  Bskurt = list()
  
  SBsmean = list() 
  SBsvar  = list()
  SBsskew = list()
  SBskurt = list()
  
  APEsmean = list() 
  APEsvar  = list()
  APEsskew = list()
  APEskurt = list()
  
  etimes = list()
  
  
  for(i in 1:NRuns){
    
    Bsmean[[i]] = do.call(rbind, lapply(Results[[i]], `[`, 1,))
    Bsvar[[i]]  = do.call(rbind, lapply(Results[[i]], `[`, 2,))
    Bsskew[[i]] = do.call(rbind, lapply(Results[[i]], `[`, 3,))
    Bskurt[[i]] = do.call(rbind, lapply(Results[[i]], `[`, 4,))
    
    SBsmean[[i]] = Bsmean[[i]]^2
    SBsvar[[i]]  = Bsvar[[i]]^2
    SBsskew[[i]] = Bsskew[[i]]^2
    SBskurt[[i]] = Bskurt[[i]]^2
    
    APEsmean[[i]] = do.call(rbind, lapply(Results[[i]], `[`, 5,))
    APEsvar[[i]]  = do.call(rbind, lapply(Results[[i]], `[`, 6,))
    APEsskew[[i]] = do.call(rbind, lapply(Results[[i]], `[`, 7,))
    APEskurt[[i]] = do.call(rbind, lapply(Results[[i]], `[`, 8,))
    
    etimes[[i]]   = do.call(rbind, lapply(Results[[i]], `[`, 9,))

  } 

  
  
  # Calculate several error measures
  
  
  MAEmean =  Reduce('+', lapply(Bsmean, abs)) / NRuns
  MAEvar =  Reduce('+', lapply(Bsvar, abs)) / NRuns
  MAEskew =  Reduce('+', lapply(Bsskew, abs)) / NRuns
  MAEkurt =  Reduce('+', lapply(Bskurt, abs)) / NRuns
  
  
  RMSEmean = sqrt(Reduce('+', SBsmean) / NRuns) 
  RMSEvar  = sqrt(Reduce('+', SBsvar) / NRuns)
  RMSEskew = sqrt(Reduce('+', SBsskew) / NRuns)
  RMSEkurt = sqrt(Reduce('+', SBskurt) / NRuns)
  
  
  MAPEmean = Reduce('+', APEsmean) / NRuns # Alternatively: MAPEmean = apply(simplify2array(APEsmean), 1:2, mean)
  MAPEvar  = Reduce('+', APEsvar) / NRuns
  MAPEskew = Reduce('+', APEsskew) / NRuns
  MAPEkurt = Reduce('+', APEskurt) / NRuns
  
  # Calculate mean elapsed time
  
  Metime = Reduce('+', etimes) / NRuns
  
  # varAPEsmean = apply(simplify2array(APEsmean), 1:2, var)
  
  # Calculate respective confidence bounds
  
  q05AEsmean = apply(simplify2array(lapply(Bsmean, abs)), 1:2, quantile, 0.05)
  q95AEsmean = apply(simplify2array(lapply(Bsmean, abs)), 1:2, quantile, 0.95)
  
  q05AEsvar = apply(simplify2array(lapply(Bsvar, abs)), 1:2, quantile, 0.05)
  q95AEsvar = apply(simplify2array(lapply(Bsvar, abs)), 1:2, quantile, 0.95)
  
  q05AEsskew = apply(simplify2array(lapply(Bsskew, abs)), 1:2, quantile, 0.05)
  q95AEsskew = apply(simplify2array(lapply(Bsskew, abs)), 1:2, quantile, 0.95)
  
  q05AEskurt = apply(simplify2array(lapply(Bskurt, abs)), 1:2, quantile, 0.05)
  q95AEskurt = apply(simplify2array(lapply(Bskurt, abs)), 1:2, quantile, 0.95)
  
  q05Bsmean = sqrt(apply(simplify2array(SBsmean), 1:2, quantile, 0.05))
  q95Bsmean = sqrt(apply(simplify2array(SBsmean), 1:2, quantile, 0.95))
  
  q05Bsvar = sqrt(apply(simplify2array(SBsvar), 1:2, quantile, 0.05))
  q95Bsvar = sqrt(apply(simplify2array(SBsvar), 1:2, quantile, 0.95))
  
  q05Bsskew = sqrt(apply(simplify2array(SBsskew), 1:2, quantile, 0.05))
  q95Bsskew = sqrt(apply(simplify2array(APEsskew), 1:2, quantile, 0.95))
  
  q05Bskurt = sqrt(apply(simplify2array(SBskurt), 1:2, quantile, 0.05))
  q95Bskurt = sqrt(apply(simplify2array(SBskurt), 1:2, quantile, 0.95))
  
  q05APEsmean = apply(simplify2array(APEsmean), 1:2, quantile, 0.05)
  q95APEsmean = apply(simplify2array(APEsmean), 1:2, quantile, 0.95)
  
  q05APEsvar = apply(simplify2array(APEsvar), 1:2, quantile, 0.05)
  q95APEsvar = apply(simplify2array(APEsvar), 1:2, quantile, 0.95)
  
  q05APEsskew = apply(simplify2array(APEsskew), 1:2, quantile, 0.05)
  q95APEsskew = apply(simplify2array(APEsskew), 1:2, quantile, 0.95)
  
  q05APEskurt = apply(simplify2array(APEskurt), 1:2, quantile, 0.05)
  q95APEskurt = apply(simplify2array(APEskurt), 1:2, quantile, 0.95)
  
  Moments = factor(c("Mean", "Variance", "Skewness", "Kurtosis"), levels = c("Mean", "Variance", "Skewness", "Kurtosis"))
  groups =  factor(c("Explicit", "Implicit", "Others"), levels = c("Explicit", "Implicit", "Others")) 

  Delta = T / n
  
  df.MErrors = data.frame("Scheme" = rep(rep(names, each = length(Delta)), length(Moments)), "Group" = rep(rep(groups, c(nE * length(Delta), nI * length(Delta), nO * length(Delta))), length(Moments)), 
                          "Delta" = rep(rep(Delta, length(names)), length(Moments)), "Moment" = rep(Moments, each = length(names) * length(Delta)), 
                          "MAE" = c(c(MAEmean), c(MAEvar), c(MAEskew), c(MAEkurt)), "q05AE" = c(c(q05AEsmean), c(q05AEsvar), c(q05AEsskew), c(q05AEskurt)),
                          "q95AE" = c(c(q95AEsmean), c(q95AEsvar), c(q95AEsskew), c(q95AEskurt)),
                          "RMSE" = c(c(RMSEmean), c(RMSEvar), c(RMSEskew), c(RMSEkurt)), "q05RSE" = c(c(q05Bsmean), c(q05Bsvar), c(q05Bsskew), c(q05Bskurt)),
                          "q95RSE" = c(c(q95Bsmean), c(q95Bsvar), c(q95Bsskew), c(q95Bskurt)),
                          "MAPE" = c(c(MAPEmean), c(MAPEvar), c(MAPEskew), c(MAPEkurt)), "q05APE" = c(c(q05APEsmean), c(q05APEsvar), c(q05APEsskew), c(q05APEskurt)),
                          "q95APE" = c(c(q95APEsmean), c(q95APEsvar), c(q95APEsskew), c(q95APEskurt)),
                          "Mean Time" = rep(c(Metime), length(Moments)))
  
  return(df.MErrors)
  
}

File: Matlab/CIR/getMErrors.m
function [MAEmean, MAEvar, MAEskew, MAEkurt, RMSEmean, RMSEvar, RMSEskew, RMSEkurt, MAPEmean, MAPEvar, MAPEskew, MAPEkurt, Metime] = getMErrors(v0, kappa, theta, sigma, n, T, M, NRuns, VarM, setup) 

if strcmp(setup, 'C')
    
    mth = MomentsCIR(1:4, 0, kappa, theta, sigma, v0, T, 1, 0);
    vthV = mth(2) - mth(1)^2;
    skthV= getSkFromMoments(mth(1), mth(2), mth(3));
    kuthV= getKuFromMoments(mth(1), mth(2), mth(3), mth(4));
    
    momth.m = mth(1); momth.v = vthV; momth.sk = skthV; momth.ku = kuthV;

    % feature('numcores'); % Edit the number of used cores (NumWorkers) in Cluster profile 'Local' if it is not the number of all logical cores matlab is using
    parpool(size(VarM,1));

    % since we are making many steps in each algo, I think it would be better
    % to parallelize the nested for-loop instead of the bigger one (here, it is
    % therefore exactly the other way around compared to the "unconditional"
    % setting)

    for k=1:NRuns

        for j=1:size(n,2)

            parfor i = 1:size(VarM,1)

                Stats(:,i) = TreatVariance(VarM{i}, M, n(j), T, kappa, theta, sigma, v0, momth, setup);

            end

            Errors{j} = Stats;

        end

        Res{k} = Errors;    
        disp(k)

    end

    delete(gcp('nocreate'));
    
    for k=1:NRuns 

       for j = 1:length(n)

          auxBsmean(j,:) = Res{k}{j}(1,:);
          auxBsvar(j,:) = Res{k}{j}(2,:);
          auxBsskew(j,:) = Res{k}{j}(3,:);
          auxBskurt(j,:) = Res{k}{j}(4,:);

          auxSBsmean = auxBsmean.^2;
          auxSBsvar  =  auxBsvar.^2;
          auxSBsskew = auxBsskew.^2;
          auxSBskurt = auxBskurt.^2;

          auxAPEsmean(j,:) = Res{k}{j}(5,:);
          auxAPEsvar(j,:) = Res{k}{j}(6,:);
          auxAPEsskew(j,:) = Res{k}{j}(7,:);
          auxAPEskurt(j,:) = Res{k}{j}(8,:);

          auxTimes(j,:) = Res{k}{j}(9,:);

        end    

        Bsmean{k} = auxBsmean;
        Bsvar{k} = auxBsvar;
        Bsskew{k} = auxBsskew;
        Bskurt{k} = auxBskurt;

        SBsmean{k} = auxSBsmean;
        SBsvar{k}  = auxSBsvar;
        SBsskew{k} = auxSBsskew;
        SBskurt{k} = auxSBskurt;

        APEsmean{k} = auxAPEsmean;
        APEsvar{k} = auxAPEsvar;
        APEsskew{k} = auxAPEsskew;
        APEskurt{k} = auxAPEskurt;

        times{k} = auxTimes;

    end

elseif strcmp(setup, 'UC')
    
    % plug in some dummy for T and v0 here, since the unconditional
    % moments of CIR are not function of these parameters
    mth = MomentsCIR(1:4, 0, kappa, theta, sigma, 1, 1, 0, 0);
    vthV = mth(2) - mth(1)^2;
    skthV= getSkFromMoments(mth(1), mth(2), mth(3));
    kuthV= getKuFromMoments(mth(1), mth(2), mth(3), mth(4));
    
    momth.m = mth(1); momth.v = vthV; momth.sk = skthV; momth.ku = kuthV;
    
    parpool(size(VarM,1));
    
     for k=1:NRuns

        for j=1:size(T,2)
 
            parfor i = 1:size(VarM,1)
                    
                    % To set setup = 'C' here is not a mistake, since we are interested in the distribution of v_T again
                    Stats(:,i) = TreatVariance(VarM{i}, M, n, T(j), kappa, theta, sigma, v0, momth, 'C');

             end

         Errors{j} = Stats;

        end

        Res{k} = Errors;    
        disp(k)

    end

    delete(gcp('nocreate'));
    
    for k=1:NRuns 

       for j = 1:length(T)

          auxBsmean(j,:) = Res{k}{j}(1,:);
          auxBsvar(j,:) = Res{k}{j}(2,:);
          auxBsskew(j,:) = Res{k}{j}(3,:);
          auxBskurt(j,:) = Res{k}{j}(4,:);

          auxSBsmean = auxBsmean.^2;
          auxSBsvar  =  auxBsvar.^2;
          auxSBsskew = auxBsskew.^2;
          auxSBskurt = auxBskurt.^2;

          auxAPEsmean(j,:) = Res{k}{j}(5,:);
          auxAPEsvar(j,:) = Res{k}{j}(6,:);
          auxAPEsskew(j,:) = Res{k}{j}(7,:);
          auxAPEskurt(j,:) = Res{k}{j}(8,:);

          auxTimes(j,:) = Res{k}{j}(9,:);

        end    

        Bsmean{k} = auxBsmean;
        Bsvar{k} = auxBsvar;
        Bsskew{k} = auxBsskew;
        Bskurt{k} = auxBskurt;

        SBsmean{k} = auxSBsmean;
        SBsvar{k}  = auxSBsvar;
        SBsskew{k} = auxSBsskew;
        SBskurt{k} = auxSBskurt;

        APEsmean{k} = auxAPEsmean;
        APEsvar{k} = auxAPEsvar;
        APEsskew{k} = auxAPEsskew;
        APEskurt{k} = auxAPEskurt;

        times{k} = auxTimes;

    end
    
    
end 


MAEmean = sum(abs(cat(3,Bsmean{:})),3) / NRuns;
MAEvar = sum(abs(cat(3,Bsvar{:})),3) / NRuns;
MAEskew = sum(abs(cat(3,Bsskew{:})),3) / NRuns;
MAEkurt = sum(abs(cat(3,Bskurt{:})),3) / NRuns;

RMSEmean = sqrt(sum(cat(3,SBsmean{:}),3) / NRuns);
RMSEvar = sqrt(sum(cat(3,SBsvar{:}),3) / NRuns);
RMSEskew = sqrt(sum(cat(3,SBsskew{:}),3) / NRuns);
RMSEkurt = sqrt(sum(cat(3,SBskurt{:}),3) / NRuns);

MAPEmean = sum(cat(3,APEsmean{:}),3) / NRuns;
MAPEvar = sum(cat(3,APEsvar{:}),3) / NRuns;
MAPEskew = sum(cat(3,APEsskew{:}),3) / NRuns;
MAPEkurt = sum(cat(3,APEskurt{:}),3) / NRuns;

Metime =  sum(cat(3,times{:}),3) / NRuns;

